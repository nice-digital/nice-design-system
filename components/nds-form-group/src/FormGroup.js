import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import "./../scss/form-group.scss";

export const FormGroup = props => {
	const { groupError, inline, legend, children, hint, name, ...rest } = props;

	const clonedChildren = React.Children.map(children, child => {
		return React.cloneElement(child, {
			name,
			inline,
			...rest
		});
	});

	const classes = classnames({
		"form-group": true,
		"form-group--no-legend": legend ? false : true
	});

	return (
		<fieldset className={classes}>
			{legend && <legend className="form-group__legend">{legend}</legend>}
			{groupError && <p className="form-group__error-message">{groupError}</p>}
			{hint && <p className="form-group__hint">{hint}</p>}
			{clonedChildren}
		</fieldset>
	);
};

FormGroup.propTypes = {
	children: PropTypes.node.isRequired,
	legend: PropTypes.node,
	name: PropTypes.string,
	hint: PropTypes.node,
	inline: PropTypes.bool,
	groupError: PropTypes.string
};
