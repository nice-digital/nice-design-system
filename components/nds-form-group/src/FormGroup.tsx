import React from "react";
import classnames from "classnames";
import "./../scss/form-group.scss";

export interface FormGroupProps {
	[prop: string]: unknown;
	inline?: boolean;
	name?: string;
	hint?: React.ReactNode;
	legend?: string;
	children: React.ReactNode;
	groupError?: string | boolean;
}

export const FormGroup: React.FC<FormGroupProps> = (props: FormGroupProps) => {
	const { groupError, inline, legend, children, hint, name, ...rest } = props;

	const clonedChildren = React.Children.map(children, (child) => {
		return React.cloneElement(child as React.ReactElement, {
			name,
			inline,
			...rest
		});
	});

	const classes = classnames({
		"form-group": true,
		"form-group--no-legend": legend ? false : true
	});

	return (
		<fieldset
			className={classes}
			data-component={`form-group${inline ? "--inline" : ""}`}
		>
			{legend && <legend className="form-group__legend">{legend}</legend>}
			{groupError && <p className="form-group__error-message">{groupError}</p>}
			{hint && <p className="form-group__hint">{hint}</p>}
			{clonedChildren}
		</fieldset>
	);
};
