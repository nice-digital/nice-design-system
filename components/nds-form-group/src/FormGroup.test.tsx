import React from "react";
import { render } from "@testing-library/react";
import { FormGroup } from "./FormGroup";

const props = {
	legend: "This is the legend text",
	hint: "This is the hint text",
	groupError: "This is the error text",
	inline: false,
	name: "group-name"
};

describe("FormGroup", () => {
	it("should match the snapshot when no props are passed", () => {
		const wrapper = render(
			<FormGroup>
				<span>Render</span>
			</FormGroup>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match the snapshot when all possible props are passed", () => {
		const wrapper = render(
			<FormGroup {...props}>
				<span>Render</span>
			</FormGroup>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should pass any additional props down to the cloned child components", () => {
		const localProps = Object.assign({}, props);
		const { container } = render(
			<FormGroup {...localProps}>
				<span>Render</span>
			</FormGroup>
		);
		const child = container.querySelector("span");
		expect(child?.getAttribute("name")).toBe("group-name");
	});
});
