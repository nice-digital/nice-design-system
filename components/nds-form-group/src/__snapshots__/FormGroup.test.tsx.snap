// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup should match the snapshot when all possible props are passed 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <fieldset
        class="form-group"
        data-component="form-group"
      >
        <legend
          class="form-group__legend"
        >
          This is the legend text
        </legend>
        <p
          class="form-group__error-message"
        >
          This is the error text
        </p>
        <p
          class="form-group__hint"
        >
          This is the hint text
        </p>
        <span
          name="group-name"
        >
          Render
        </span>
      </fieldset>
    </div>
  </body>,
  "container": <div>
    <fieldset
      class="form-group"
      data-component="form-group"
    >
      <legend
        class="form-group__legend"
      >
        This is the legend text
      </legend>
      <p
        class="form-group__error-message"
      >
        This is the error text
      </p>
      <p
        class="form-group__hint"
      >
        This is the hint text
      </p>
      <span
        name="group-name"
      >
        Render
      </span>
    </fieldset>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`FormGroup should match the snapshot when no props are passed 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <fieldset
        class="form-group form-group--no-legend"
        data-component="form-group"
      >
        <span>
          Render
        </span>
      </fieldset>
    </div>
  </body>,
  "container": <div>
    <fieldset
      class="form-group form-group--no-legend"
      data-component="form-group"
    >
      <span>
        Render
      </span>
    </fieldset>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
